### ISSUE TITLE
Support different batch size across GPUs with DDP

### ISSUE THREAD

**Issue by maxwellzh** [pytorch/pytorch#67253]
## 🚀 Feature

Ask for DDP wrapper of `nn.Module` supports a new flag that allow different batch size across processes. So that when invoking the `loss.backward()`, DDP won't automatically do `sum(grad)/num_replicas`.
This is kind of duplicated to #31553 . But I think my case is more concrete.

## Motivation

Basically, there're two potential situations:
1. The performance across GPUs is different (e.g. Tesla V100 vs. GTX 1080Ti). We may want the GPU with better perf being fed with larger batch size.
2. Dealing with sequence problem of variable length data. To arranging variable-length data into batch, commonly we add paddings to align them to the max length. However, for DDP, this can be optimized via balancing the data on GPUs. This figure might help better explain the difference of the native ( DistributedSampler) and a optimized sampler
![batching](https://user-images.githubusercontent.com/32546978/138889118-882ec120-7700-4e23-9b12-80282576651e.png)

## Alternatives

Currently, we can do that in a hack way.
```python3
# on each process. 
# get batch size on local process 
local_batch_size = ...
# suppose the loss is reduce by 'mean', calculate the sum of loss on local process
local_loss_sum = loss.detach() * local_batch_size
# local_batch_size is reduced to global batch size
dist.all_reduce(local_batch_size)
global_batch_size = local_batch_size
# modify the loss in-place
loss.data = local_loss_sum * world_size / global_batch_size
loss.backward()
```
With this normalized loss, now in the `loss.backward()`, the DDP is actually doing 
```
sum(local_grad_sum * world_size / global_batch_size)/world_size
=global_grad_sum / global_batch_size
```
And it's the correct way of back-propagation with different batch size across GPUs.

## Additional context

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

**Comment by antopost** [comment]
I've also been thinking about training on multiple GPUs with different batch sizes.
But what exactly is the advantage of doing this?
The GPUs processing the larger batches will presumably take longer for an iteration, hence the other GPUs processing smaller batches will always be waiting at the end of each iteration before the gradient accumulation step.
So, does this really make my training faster?

**Comment by maxwellzh** [comment]
@antopost 
> The GPUs processing the larger batches will presumably take longer for an iteration

Not always. It's common in sequence modeling tasks that data in a mini-batch varies in length. 

Computation of "larger batch size * shorter seq" might be close to "smaller batch size * longer seq".

**Comment by maxwellzh** [comment]
@ritwikmishra Please have a look at this reply I once posted on pytorch forum.
https://discuss.pytorch.org/t/can-ddp-divide-dataset-unevenly-to-workers/146381/4?u=huahuanz

**Comment by ritwikmishra** [comment]
@maxwellzh thank you for the code snippet for uneven batch sizes for different GPUs while data loading. It worked fine after some customization. 

However, I have some doubts regarding your "hacky way" to calculate loss. Please help.

1. Why is it even needed? Why not simple ```loss.backward()``` and ```optim.step()```?
2. Why you ```detach()``` the loss before scaling up to obtain ```local_loss_sum```? I tried without ```detach()```, the method works fine.

**Comment by maxwellzh** [comment]
@ritwikmishra 
1. This is useful in multi-gpu DDP. Consider this case:
```
batch =4
loss = [0.1, 0.2, 0.3, 0.4]

# for single gpu training
avg_loss = sum(loss)/4 = 0.25
# cal avg_grad
avg_loss.backward()

# with multi-gpu
# a. evenly divide data into 2 GPUs
loss1 = [0.1, 0.2]
loss2 = [0.3, 0.4]
avg_loss1 = sum(loss1)/2 = 0.15
avg_loss2 = sum(loss2)/2 = 0.35

# call .backward() on individual GPUs and torch internally invokes all reducing to obtain average grad.
# the equivalent average loss would be 
avg_loss = (avg_loss1 + avg_loss2) / 2 = 0.25

# b. unevenly divide
loss1 = [0.1, 0.2, 0.3]
loss2 = [0.4]
avg_loss1 = sum(loss1)/3 = 0.2
avg_loss2 = sum(loss2)/1 = 0.4
# the equivalent average loss would be 
avg_loss = (avg_loss1 + avg_loss2) / 2 = 0.3
```

You see in `b.` the equivalent loss is not identity to that of single GPU training. That's why I take the "hack" in.

2. You're right. `.detach()` here won't make much difference.

**Comment by e-yi** [comment]
Hi @maxwellzh, I ran some tests, and it seems that modifications to `tensor.data` do not affect gradient propagation.

```
a = torch.tensor([3.]).requires_grad_()
b = torch.tensor([4.]).requires_grad_()
c = a*b+a+b

c.data = c.data/10
c.backward()
a.grad,b.grad  # (tensor([5.]), tensor([4.]))
```
```
a = torch.tensor([3.]).requires_grad_()
b = torch.tensor([4.]).requires_grad_()
c = a*b+a+b

c = c/10
c.backward()
a.grad,b.grad  # (tensor([0.5000]), tensor([0.4000]))
```

Or does DDP work in a different way?
